src/MonitorProfiler/MainProfiler/ExceptionTracker.cpp

```diff
@@ -0,0 +1,392 @@
+        classId,
+        &moduleId,
+        &typeDefToken,
+        nullptr,
+        0,
+        nullptr,
+        nullptr
+    ));
+
+    ComPtr<IMetaDataImport2> metadataImport;
+    IfFailLogRet(_corProfilerInfo->GetModuleMetaData(
+        moduleId,
+        CorOpenFlags::ofRead,
+        IID_IMetaDataImport2,
+        (IUnknown**)&metadataImport
+    ));
+
+    // CONSIDER: If generic parameters should be replaced with actual types, do that
+    // as part of calculating the full type name. Requires calling GetClassIDInfo2 to
+    // get the generic parameter ClassID values and getting their full type name.
+
+    IfFailLogRet(GetFullTypeName(metadataImport, typeDefToken, fullClassName));
+
+    return S_OK;
+}
+
+HRESULT ExceptionTracker::GetFullTypeName(IMetaDataImport2* metadataImport, mdTypeDef typeDefToken, tstring& fullTypeName)
+{
+    HRESULT hr = S_OK;
+
+    ULONG typeNameCount = 0; // Includes null-terminater
+    DWORD typeDefFlags = 0;
+    IfFailLogRet(metadataImport->GetTypeDefProps(
+        typeDefToken,
+        nullptr,
+        0,
+        &typeNameCount,
+        &typeDefFlags,
+        nullptr
+    ));
+    
+    tstring parentTypeName;
+    if (IsNestedType(static_cast<CorTypeAttr>(typeDefFlags)))
+    {
+        mdTypeDef parentTypeDefToken;
+        IfFailLogRet(metadataImport->GetNestedClassProps(typeDefToken, &parentTypeDefToken));
+
+        IfFailLogRet(GetFullTypeName(metadataImport, parentTypeDefToken, parentTypeName));
+    }
+
+    unique_ptr<WCHAR[]> className(new (nothrow) WCHAR[typeNameCount]);
+    IfNullRet(className);
+
+    IfFailLogRet(metadataImport->GetTypeDefProps(
+        typeDefToken,
+        className.get(),
+        typeNameCount,
+        nullptr,
+        nullptr,
+        nullptr
+    ));
+
+    // The full type name should be in the following format: <OuterTypeNamespace>.<OuterType>+<InnerType>
+    // Example: System.Collections.Generic.Dictionary`2+Enumerator
+    fullTypeName.clear();
+    if (parentTypeName.length() > 0)
+    {
+        fullTypeName.append(parentTypeName);
+        fullTypeName.append(_T("+"));
+    }
+    fullTypeName.append(className.get());
+
+    return S_OK;
+}
+
+bool ExceptionTracker::IsNestedType(CorTypeAttr attributes)
+{
+    CorTypeAttr visibility = static_cast<CorTypeAttr>(attributes & CorTypeAttr::tdVisibilityMask);
+    return CorTypeAttr::tdNestedPublic <= visibility && visibility <= CorTypeAttr::tdNestedFamORAssem;
+}
+
+HRESULT ExceptionTracker::LogExceptionThrownFrame(FunctionID functionId, COR_PRF_FRAME_INFO frameInfo)
+{
+    HRESULT hr = S_OK;
+
+    tstring methodName;
+    IfFailLogRet(GetFullyQualifiedMethodName(functionId, frameInfo, methodName));
+    LogDebugV(_T("Exception thrown: %s"), methodName.c_str());
+
+    return S_OK;
+}
+
+HRESULT ExceptionTracker::LogExceptionThrownFrameCallback(
+    FunctionID funcId,
+    UINT_PTR ip,
+    COR_PRF_FRAME_INFO frameInfo,
+    ULONG32 contextSize,
```

Comment:
What are `ip` and `contextSize` doing?

Reply:
This signature is `DoStackSnapshot`'s callback, by setting various flags you can optionally get the register context for each frame, it's expensive and shouldn't be necessary here so it's safe to ignore. Without requesting the context `contextSize` and `context` should be default values.

`ip` is always passed since we have the information, it's cheap to provide and a lot of stack walkers want it, but `funcId` and `frameInfo` are sufficient to uniquely identify a method without `ip`.